core/components/addressable_pq.hpp

```diff
@@ -0,0 +1,199 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_COMPONENTS_ADDRESSABLE_PQ_HPP_
+#define GKO_CORE_COMPONENTS_ADDRESSABLE_PQ_HPP_
+
+
+#include <algorithm>
+#include <vector>
+
+
+#include <ginkgo/core/base/types.hpp>
+
+
+namespace gko {
+
+
+/**
+ * An addressable priority queue based on a k-ary heap.
+ *
+ * It allows inserting key-value pairs, modifying their key as well as accessing
+ * and removing the key-value pair with the minimum key.
+ *
+ * @tparam KeyType    The type of the keys
+ * @tparam ValueType  The type of the values
+ */
+template <typename KeyType, typename ValueType>
+struct addressable_priority_queue {
+    explicit addressable_priority_queue(int deg_log2) : degree{1 << deg_log2} {}
+
+    /**
+     * Inserts the given key-value pair into the PQ.
+     * Duplicate keys are allowed, they may be returned in an arbitrary order.
+     * @returns a handle for the pair to be used when modifying the key.
+     */
+    std::size_t insert(KeyType key, ValueType value)
+    {
+        m_keys.push_back(key);
+        m_values.push_back(value);
+        auto handle = next_handle();
+        m_handles.push_back(handle);
+        if (handle == m_handle_pos.size())
+            m_handle_pos.push_back(size() - 1);
+        else
+            m_handle_pos[handle] = size() - 1;
+        sift_up(size() - 1);
+        return handle;
+    }
+
+    /**
+     * Updates the key of the pair with the given handle.
+     */
+    void update_key(std::size_t handle, KeyType new_key)
+    {
+        auto pos = m_handle_pos[handle];
+        GKO_ASSERT(pos < size());
+        GKO_ASSERT(m_handles[pos] == handle);
+        auto old_key = m_keys[pos];
+        m_keys[pos] = new_key;
+        if (old_key < new_key) {
+            sift_down(pos);
+        } else {
+            sift_up(pos);
+        }
+    }
+
+    /**
+     * Returns the minimum key from the queue.
+     */
+    KeyType min_key() const { return m_keys[0]; }
+
+    /**
+     * Returns the value belonging to the minimum key from the queue.
+     */
+    ValueType min_val() const { return m_values[0]; }
+
+    /**
+     * Returns the key-value pair with the minimum key from the queue.
+     */
+    std::pair<KeyType, ValueType> min() const { return {min_key(), min_val()}; }
+
+    /**
+     * Removes the key-value pair with the minimum key from the queue.
+     */
+    void pop_min()
+    {
+        swap(0, size() - 1);
+        m_keys.pop_back();
+        m_values.pop_back();
+        auto old_handle = m_handles.back();
+        m_handles.pop_back();
+        sift_down(0);
+    }
+
+    /**
+     * Returns the number of key-value pairs in the queue.
+     */
```

Comment:
For consistency  I would add `@return` even if trivial
```suggestion
     *
     *  @return  the number of key-value pairs in the queue
     */
   ```